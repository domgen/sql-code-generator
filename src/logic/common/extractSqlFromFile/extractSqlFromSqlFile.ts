import { readFileAsync } from '../readFileAsync';

export const extractSqlFromSqlFile = ({ filePath }: { filePath: string }) => readFileAsync({ filePath }); // the sql is everything in the file in this case
