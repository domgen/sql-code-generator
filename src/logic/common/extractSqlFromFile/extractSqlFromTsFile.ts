import { readFileAsync } from '../readFileAsync';

/*
  note: in order to not block users when they have typescript type errors (by using ts-node),
    we can simply parse the file contents and grab the `export const sql = `__SQL__`;` w/ regex
*/
export const extractSqlFromTsFile = async ({ filePath }: { filePath: string }) => {
  // grab file contents
  const content = await readFileAsync({ filePath });

  // grab the sql definition w/ regex
  const [
    _, // tslint:disable-line no-unused
    sqlMatch,
  ] = new RegExp(/(?:export const sql = `)((.|\n)+)(?:`)/).exec(content) ?? [];

  // throw error if could not extract
  if (!sqlMatch) throw new Error(`could not extract sql from file at path '${filePath}'`);

  // return the sql
  return sqlMatch;
};
