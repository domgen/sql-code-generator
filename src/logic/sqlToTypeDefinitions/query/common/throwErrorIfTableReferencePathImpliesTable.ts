/**
 * instead of "select id from user" we're asking the user to "select user.id from user" or "select u.id from user u"
 *
 * rationale:
 * - this makes it so that we don't have to try and figure out the implicit reference
 * - this makes it so that when a user does join to another table, they don't have to add the explicit references after the fact (i.e., makes it easier to extend their queries)
 */
export const throwErrorIfTableReferencePathImpliesTable = ({ referencePath }: { referencePath: string }) => {
  if (referencePath.split('.').length !== 2) {
    throw new Error(
      `table reference of "${referencePath}" does not have the table alias explicitly defined, violating best practice`,
    );
  }
};
