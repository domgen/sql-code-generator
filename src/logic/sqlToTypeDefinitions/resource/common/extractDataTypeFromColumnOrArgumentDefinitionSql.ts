import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (mysqlStringTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.STRING;
  if (mysqlNumberTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.NUMBER;
  if (mysqlDateTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.DATE;
  if (mysqlBinaryTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
