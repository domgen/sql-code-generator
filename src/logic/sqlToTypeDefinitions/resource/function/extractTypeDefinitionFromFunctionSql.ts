import { extractInputsFromFunctionSql } from './extractInputsFromFunctionSql/extractInputsFromFunctionSql';
import { extractOutputFromFunctionSql } from './extractOutputFromFunctionSql/extractOutputFromFunctionSql';
import { TypeDefinitionOfResourceFunction } from '../../../../domain/objects/TypeDefinitionOfResourceFunction';

export const extractTypeDefinitionFromFunctionSql = ({
  name,
  sql,
}: {
  name: string;
  sql: string;
}) => {
  // 1. get the input definition
  const inputs = extractInputsFromFunctionSql({ sql });

  // 2. get the output definition
  const output = extractOutputFromFunctionSql({ sql });

  // 3. return the full type
  return new TypeDefinitionOfResourceFunction({
    name,
    inputs,
    output,
  });
};
