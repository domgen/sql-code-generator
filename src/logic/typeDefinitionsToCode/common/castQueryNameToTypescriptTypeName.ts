import { pascalCase } from 'pascal-case';

export const castQueryNameToTypescriptTypeName = ({
  name,
}: {
  name: string;
}) => {
  const queryNameInPascalCase = pascalCase(name);
  return `SqlQuery${queryNameInPascalCase}`;
};
