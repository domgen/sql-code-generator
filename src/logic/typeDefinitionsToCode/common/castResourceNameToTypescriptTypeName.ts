import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../model';

export const castResourceNameToTypescriptTypeName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
