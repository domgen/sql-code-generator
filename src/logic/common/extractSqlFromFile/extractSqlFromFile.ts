import { extractSqlFromSqlFile } from './extractSqlFromSqlFile';
import { extractSqlFromTsFile } from './extractSqlFromTsFile';

const SUPPORTED_FILE_EXTENSIONS = ['.sql', '.ts'];
export const extractSqlFromFile = async ({
  filePath,
}: {
  filePath: string;
}) => {
  // 1. grab extension from file; check that we support it
  const extension = `.${filePath.split('.').slice(-1)[0]}`;
  if (!SUPPORTED_FILE_EXTENSIONS.includes(extension)) {
    throw new Error(`file extension '.${extension}' is not supported`);
  }

  // 2. grab sql from file, based on file type
  if (extension === '.sql') return extractSqlFromSqlFile({ filePath });
  if (extension === '.ts') return extractSqlFromTsFile({ filePath });
  throw new Error('unexpected'); // should never reach here, since we checked earlier that the extension was supported
};
