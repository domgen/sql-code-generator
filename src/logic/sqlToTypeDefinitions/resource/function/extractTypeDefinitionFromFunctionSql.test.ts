import { extractSqlFromFile } from '../../../common/extractSqlFromFile';
import { extractTypeDefinitionFromFunctionSql } from './extractTypeDefinitionFromFunctionSql';

describe('extractTypeDefinitionFromFunctionSql', () => {
  it('should be able to extract types in this example', async () => {
    const exampleSql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/functions/upsert_image.sql`,
    });
    const typeDef = extractTypeDefinitionFromFunctionSql({ name: 'upsert_image', sql: exampleSql });
    expect(typeDef).toMatchSnapshot();
  });
  it('should be able to extract types in this other example', async () => {
    const exampleSql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/functions/upsert_suggestion.sql`,
    });
    const typeDef = extractTypeDefinitionFromFunctionSql({ name: 'upsert_suggestion', sql: exampleSql });
    expect(typeDef).toMatchSnapshot();
  });
  it('should be able to extract types in this other other example', async () => {
    const exampleSql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/functions/hash_string.sql`,
    });
    const typeDef = extractTypeDefinitionFromFunctionSql({ name: 'hash_string', sql: exampleSql });
    expect(typeDef).toMatchSnapshot();
  });
});
