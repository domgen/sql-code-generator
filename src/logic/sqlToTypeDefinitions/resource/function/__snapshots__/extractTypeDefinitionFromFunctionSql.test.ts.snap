// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTypeDefinitionFromFunctionSql mysql should be able to extract types in this example 1`] = `
TypeDefinitionOfResourceFunction {
  "inputs": [
    TypeDefinitionOfResourceInput {
      "name": "in_url",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_caption",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_credit",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_alt_text",
      "type": [
        "string",
        "null",
      ],
    },
  ],
  "name": "upsert_image",
  "output": [
    "number",
  ],
}
`;

exports[`extractTypeDefinitionFromFunctionSql mysql should be able to extract types in this other example 1`] = `
TypeDefinitionOfResourceFunction {
  "inputs": [
    TypeDefinitionOfResourceInput {
      "name": "in_suggestion_source",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_external_id",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_suggested_idea_id",
      "type": [
        "number",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_status",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_result",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_resultant_curated_idea_id",
      "type": [
        "number",
        "null",
      ],
    },
  ],
  "name": "upsert_suggestion",
  "output": [
    "number",
  ],
}
`;

exports[`extractTypeDefinitionFromFunctionSql mysql should be able to extract types in this other other example 1`] = `
TypeDefinitionOfResourceFunction {
  "inputs": [
    TypeDefinitionOfResourceInput {
      "name": "in_array",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_delimiter",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_index",
      "type": [
        "number",
        "null",
      ],
    },
  ],
  "name": "hash_string",
  "output": [
    "string",
  ],
}
`;

exports[`extractTypeDefinitionFromFunctionSql postgres should be able to extract types from a function that returns a table 1`] = `
TypeDefinitionOfResourceFunction {
  "inputs": [
    TypeDefinitionOfResourceInput {
      "name": "in_uuid",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_title",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_status",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_description",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_location_id",
      "type": [
        "number",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_photo_ids",
      "type": [
        "number[]",
        "null",
      ],
    },
  ],
  "name": "upsert_jerb",
  "output": TypeDefinitionOfResourceTable {
    "columns": [
      TypeDefinitionOfResourceColumn {
        "name": "id",
        "type": [
          "number",
        ],
      },
      TypeDefinitionOfResourceColumn {
        "name": "uuid",
        "type": [
          "string",
        ],
      },
      TypeDefinitionOfResourceColumn {
        "name": "created_at",
        "type": [
          "Date",
        ],
      },
      TypeDefinitionOfResourceColumn {
        "name": "effective_at",
        "type": [
          "Date",
        ],
      },
    ],
    "name": "function.output",
  },
}
`;

exports[`extractTypeDefinitionFromFunctionSql postgres should be able to extract types in this example 1`] = `
TypeDefinitionOfResourceFunction {
  "inputs": [
    TypeDefinitionOfResourceInput {
      "name": "in_url",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_description",
      "type": [
        "string",
        "null",
      ],
    },
  ],
  "name": "upsert_photo",
  "output": [
    "number",
  ],
}
`;

exports[`extractTypeDefinitionFromFunctionSql postgres should be able to extract types in this other example 1`] = `
TypeDefinitionOfResourceFunction {
  "inputs": [
    TypeDefinitionOfResourceInput {
      "name": "in_uuid",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_title",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_status",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_description",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_location_id",
      "type": [
        "number",
        "null",
      ],
    },
    TypeDefinitionOfResourceInput {
      "name": "in_photo_ids",
      "type": [
        "number[]",
        "null",
      ],
    },
  ],
  "name": "upsert_job",
  "output": [
    "number",
  ],
}
`;
